src/platform/w3dengine/client/w3dterrainvisual.cpp

```diff
@@ -0,0 +1,628 @@
+
+        points[1].X = major_radius + exit_width + bib_width;
+        points[1].Y = minor_radius + bib_width;
+        points[1].Z = 0.0f;
+
+        points[2].X = points[1].X;
+        points[2].Y = minor_radius + bib_width;
+        points[2].Z = 0.0f;
+
+        points[3].X = points[0].X;
+        points[3].Y = minor_radius + bib_width;
+        points[3].Z = 0.0f;
+
+        Matrix3D::Transform_Vector(*drawable->Get_Transform_Matrix(), points[0], &points[0]);
+        Matrix3D::Transform_Vector(*drawable->Get_Transform_Matrix(), points[1], &points[1]);
+        Matrix3D::Transform_Vector(*drawable->Get_Transform_Matrix(), points[2], &points[2]);
+        Matrix3D::Transform_Vector(*drawable->Get_Transform_Matrix(), points[3], &points[3]);
+
+        m_baseHeightMap->Add_Terrain_Bib_Drawable(points, drawable->Get_ID(), highlighted);
+    }
+}
+
+void W3DTerrainVisual::Remove_Faction_Bib_Drawable(Drawable *drawable)
+{
+    if (m_baseHeightMap != nullptr) {
+        m_baseHeightMap->Remove_Terrain_Bib_Drawable(drawable->Get_ID());
+    }
+}
+
+void W3DTerrainVisual::Remove_All_Bibs()
+{
+    if (m_baseHeightMap != nullptr) {
+        m_baseHeightMap->Remove_All_Terrain_Bibs();
+    }
+}
+
+void W3DTerrainVisual::Remove_Bib_Highlighting()
+{
+    if (m_baseHeightMap != nullptr) {
+        m_baseHeightMap->Remove_Terrain_Bib_Highlighting();
+    }
+}
+
+void W3DTerrainVisual::Remove_Trees_And_Props_For_Construction(const Coord3D *pos, const GeometryInfo &geom, float angle)
+{
+    if (m_baseHeightMap != nullptr) {
+        m_baseHeightMap->Remove_Trees_And_Props_For_Construction(pos, geom, angle);
+    }
+}
+
+void W3DTerrainVisual::Add_Prop(const ThingTemplate *tmplate, const Coord3D *pos, float angle)
+{
+    BitFlags<MODELCONDITION_COUNT> flags;
+
+    if (g_theWriteableGlobalData->m_weather == WEATHER_SNOWY) {
+        flags.Set(MODELCONDITION_SNOW, true);
+    }
+
+    if (g_theWriteableGlobalData->m_timeOfDay == TIME_OF_DAY_NIGHT) {
+        flags.Set(MODELCONDITION_NIGHT, true);
+    }
+
+    Utf8String name;
+    float scale = tmplate->Get_Asset_Scale();
+    const ModuleData *data = tmplate->Get_Draw_Modules()->Get_Nth_Data(0);
+
+    if (data != nullptr) {
+        const W3DModelDrawModuleData *draw_data = data->Get_As_W3D_Model_Draw_Module_Data();
+
+        if (draw_data != nullptr) {
+            name = draw_data->Get_Best_Model_Name_For_WB(flags);
+        }
+    }
+
+    if (m_baseHeightMap != nullptr && name.Is_Not_Empty()) {
+        m_baseHeightMap->Add_Prop(1, *pos, angle, scale, name);
+    }
+}
+
+void W3DTerrainVisual::Set_Raw_Map_Height(const ICoord2D *pos, int height)
+{
+    if (m_heightMap != nullptr) {
+        int x = m_heightMap->Get_Border_Size() + pos->x;
+        int y = m_heightMap->Get_Border_Size() + pos->y;
+        int map_height = m_heightMap->Get_Height(x, y);
+
+        if (map_height > height) {
+            m_heightMap->Set_Height(x, y, height);
+            m_baseHeightMap->Static_Lighting_Changed();
```

Comment:
There is no test for `m_baseHeightMap != nullptr` in this function.

Reply:
Original has no check.